platform-sdk/docs/proposals/virtualmegamap-binary-merkledb/proposal.md

```diff
@@ -0,0 +1,162 @@
+
+* `loadLeafRecord()` now returns a `VirtualLeafBytes` (was: `VirtualLeafRecord`)
+* `loadHash()` now returns a `VirtualHashBytes` (was: `VirtualHashRecord`)
+* `saveRecords()` now accepts streams of `VirtualLeafBytes` and `VirtualHashBytes` (was: streams of
+  records)
+
+`VirtualLeafBytes` is a binary counterpart of the existing `VirtualLeafRecord`. It also contains a
+path, a key, and a value, but the key and the value are immutable `Bytes` (PBJ type), not typed
+`VirtualKey` and `VirtualValue` subclasses.
+
+`VirtualHashBytes` is a similar concept for virtual internal nodes that store hashes.
+
+### Serialization
+
+Virtual leaf and hash records are serialized to bytes at the virtual map level before they get
+flushed to disk:
+
+* Copy flushes: in `VirtualRootNode`
+* Reconnect flushes: in `AbstractHashListener`
+
+Serialization is done using `KeySerializer` and `ValueSerializer`. These interfaces are moved from
+MerkleDb to VirtualMap Java modules. Key and value serializer instances are now provided by the app
+to virtual maps, not to MerkleDb data sources.
+
+### Deserialization
+
+All keys and values are stored as untyped bytes objects on disk. When they are read from disk,
+MerkleDb returns them as bytes (`VirtualLeafBytes` or `VirtualHashBytes` objects) to the caller,
+usually `RecordAccessorImpl`. At that level, the bytes are deserialized to Java key and value
+objects using the same serializers as in the section above.
+
+### API changes
+
+MerkleDb is a backing store for `VirtualMap`, however, it used to be accessed directly from the app
+code for one particular reason: when a virtual map was created, the app provided MerkleDb data source
+configuration to it. The configuration included key and value serializers and a few other params.
+
+Now the serializers are moved to `VirtualMap`, while the data source will not need them anymore.
+Application code has to be changed to provide the serializers to `VirtualMap` rather than to MerkleDb.
+
+`KeySerializer` and `ValueSerializer` interfaces are moved from one package to another, but all
+interface methods are the same. Application code needs to be updated to use the new package (and
+another Java module).
+
+## Compatibility and Migration
+
+No data migration is expected at this point. Data on disk is stored in the same format (100% byte to
+byte compatibility) as previously.
+
+When a new virtual map is created, it gets key and value serializers from the app. However, there
+is a number of already created maps in the old states. They are not created from scratch, but loaded
+from state snapshots. Where do they get the serializers from? MerkleDb. Despite MerkleDb no longer
+uses serializers, it still loads them from disk (MerkleDb configuration), if available. When a
+virtual map is loaded from a state snapshot, and snapshot is old (before MerkleDb is switched to
+work with bytes), the map will query the data source for key and value serializers to use.
```

Comment:
```suggestion
When a new virtual map is created, it gets key and value serializers from the app. However, there
is a number of already created maps in the old states. They are not created from scratch, but loaded
from state snapshots. Such virtual maps rely on the MerkleDb to provide the serializers. Despite MerkleDb no longer
uses serializers, it can still load them from disk (from the MerkleDb configuration), if available. When a
virtual map is loaded from a state snapshot, and the snapshot is old (before MerkleDb is switched to
work with bytes), the map will query the data source for key and value serializers to use.
```

Also, might want to expand on the `if available` piece - it's not clear when this condition is true. Is it typical or rare? Why/when would the serializers be missing from the MerkleDb configuration in the current state?

Also, it's not clear whether the serializers read from the MerkleDb configuration would override the ones specified in the data source or the latter take the precedence. If the latter is true, and I presume a code change would update all the data sources to specify the serializers when this change is rolled out, then it's unclear why we even care about the serializers stored in the MerkleDb configuration at all. It would be nice to clarify these points.

Reply:
"If available" here means "when the snapshot was taken before this feature". In the current `develop`, configuration of any MerkleDb table contains key and value serializers. They are stored in MerkleDb metadata file, `database_metadata.pbj` After this feature is merged, the serializers will not be there.

So the code will be like this:
* When a data source is loaded from a snapshot, it reads its table config from MerkleDb metadata. If key / value serializers are there, they are loaded. If not - no error
* When a virtual map is loaded from a snapshot, it checks its serialization version. Based on the version, it will either deserialize the serializers or not
* If the version is "old", and no serializers are available, the virtual map will get them from the data source, it's guaranteed to have them because of the version (if for some reason there are no serializers, an exception is thrown)
* When a virtual map is created from scratch, it will have the serializers set explicitly, it won't try to load them from the data source

So the short answer is: "if available" means an old version, before this feature is implemented. When a snapshot is taken with a new version, the serializers will not be in the data source, nor VirtualMap will query the data source for them.

I am not sure how to describe it in clear words in the proposal...

Reply:
Thanks for applying the change, it looks great.

And thanks for providing the extra details about the "if available" part. I think you could put this comment almost as is in the design document to clarify the plan to deal with the serializers.

Reply:
OK, let me add it to the doc.